packages/block-library/src/page-list/edit.js

```diff
@@ -182,6 +138,10 @@ export default function PageListEdit( {
 		pages?.length > 0 &&
 		pages?.length <= MAX_PAGE_COUNT;
 
+	const [ isModalOpen, setModalOpen ] = useState( false );
+	const openModal = () => setModalOpen( true );
+	const closeModal = () => setModalOpen( false );
```

Comment:
Moving the state in here instead of its own isolated component means the whole thing will need to rerender instead of just the modal when it opens.

In this specific place, the performance impact is probably going to be negligible especially with many of the complex data fetching and processing wrapped in a `useMemo`. But I see complex components like this a lot in gutenberg that I think could be done better, so it seemed worth a note.

Reply:
I moved the state back to the child and then used a useEffect to to pass the opening function back to the parent. This avoids the rerender but I'm not sure if it's an anitpattern?

The alternative I considered was using a redux store, but we don't have a store for the block library.

Reply:
I pushed up [a commit](https://github.com/WordPress/gutenberg/pull/47922/commits/131350dc923bfc7b1e2329c5a19d775081dbec8c) with what I was trying to explain since I didn't do a very good job with English 😅 

We can't move the ToolbarButton out because then we're limited to using the Modal from the toolbar which is against the goal of this refactoring. The modal stays separate for reuse, but the open/close controls still get their own component.